File: IcoSoku/Java/src/icosoku/IcoSokuSolver.java
package icosoku;

import java.util.LinkedList;

/**
 * Created by Patrick on 22.10.2014.
 */
public class IcoSokuSolver
{
    public static boolean Solve(IcoSoku ico, int flaeche)
    {
        // Rätsel gelöst
        if (flaeche > 19)
        {
            return ico.pruefeAlles();
        }

        LinkedList<Integer> pieces = GetAvailablePieces(ico);

        // Solange Plättchen noch nicht angeschaut wurden
        while (pieces.size() > 0)
        {
            int plaettchen = pieces.get(0);

            for (int orient = 0; orient < 3; orient++)
            {
                ico.setzePlaettchen(flaeche, plaettchen, orient);
                if (ico.pruefeFlaeche(flaeche))
                {
                    // Plättchen korrekt platziert, gehe zu nächster Fläche
                    if (Solve(ico, flaeche + 1))
                    {
                        // Wenn Rätsel gelöst, verlasse Rekursion
                        return true;
                    }
                }

                // Falsch platziert oder Rätsel noch nicht gelöst
                // Entferne platziertes Plättchen
                ico.entfernePlaettchen(flaeche);

                // rotieren bei gleichen Plättchen-Zahlen abbrechen
                if (ico.istPlaettchenSymmetrisch(plaettchen))
                {
                    break;
                }
            }

            // Plättchen passt mit keiner Orientierung, entferne es und alle mit den gleichen Werten
            EntfernePlaettchen(pieces, plaettchen);
        }

        return false;
    }

    private static void EntfernePlaettchen(LinkedList<Integer> pieces, int plaettchen)
    {
        if (plaettchen == 5 || plaettchen == 6 || plaettchen == 7)
        {
            pieces.remove((Object) 5);
            pieces.remove((Object) 6);
            pieces.remove((Object) 7);

            return;
        }

        if (plaettchen == 8 || plaettchen == 9 || plaettchen == 10)
        {
            pieces.remove((Object) 8);
            pieces.remove((Object) 9);
            pieces.remove((Object) 10);

            return;
        }

        if (plaettchen == 14 || plaettchen == 15)
        {
            pieces.remove((Object) 14);
            pieces.remove((Object) 15);

            return;
        }

        if (plaettchen == 16 || plaettchen == 17)
        {
            pieces.remove((Object) 16);
            pieces.remove((Object) 17);

            return;
        }

        pieces.remove((Object) plaettchen);
    }

    private static LinkedList<Integer> GetAvailablePieces(IcoSoku ico)
    {
        LinkedList<Integer> pieces = new LinkedList<>();

        for (int plaettchen = 0; plaettchen < 20; plaettchen++)
        {
            if (ico.istPlaettchenVerfuegbar(plaettchen))
            {
                pieces.add(plaettchen);
            }
        }

        return pieces;
    }
}

File: IcoSoku/C#/IcoSokuSolver.cs
using System.Collections.Generic;
using System.Linq;

namespace IcoSoku
{
    public class IcoSokuSolver
    {
        public static bool Solve(IcoSoku ico, int flaeche)
        {
            // Rätsel gelöst
            if (flaeche > 19)
            {
                return ico.PruefeAlles();
            }

            var pieces = GetAvailablePieces(ico);

            // Solange Plättchen noch nicht angeschaut wurden

            while (pieces.Count > 0)
            {
                var plaettchen = pieces.First();

                for (var orient = 0; orient < 3; orient++)
                {
                    ico.SetzePlaettchen(flaeche, plaettchen, orient);
                    if (ico.PruefeFlaeche(flaeche))
                    {
                        // Plättchen korrekt platziert, gehe zu nächster Fläche
                        if (Solve(ico, flaeche + 1))
                        {
                            // Wenn Rätsel gelöst, verlasse Rekursion
                            return true;
                        }
                    }

                    // Falsch platziert oder Rätsel noch nicht gelöst
                    // Entferne platziertes Plättchen
                    ico.EntfernePlaettchen(flaeche);

                    // rotieren bei gleichen Plättchen-Zahlen abbrechen
                    if (ico.IstPlaettchenSymmetrisch(plaettchen))
                    {
                        break;
                    }
                }

                // Plättchen passt mit keiner Orientierung, entferne es und alle mit den gleichen Werten
                EntfernePlaettchen(pieces, plaettchen);
            }

            return false;
        }

        private static void EntfernePlaettchen(LinkedList<int> pieces, int plaettchen)
        {
            if (plaettchen == 5 || plaettchen == 6 || plaettchen == 7)
            {
                pieces.Remove(5);
                pieces.Remove(6);
                pieces.Remove(7);

                return;
            }

            if (plaettchen == 8 || plaettchen == 9 || plaettchen == 10)
            {
                pieces.Remove(8);
                pieces.Remove(9);
                pieces.Remove(10);

                return;
            }

            if (plaettchen == 14 || plaettchen == 15)
            {
                pieces.Remove(14);
                pieces.Remove(15);

                return;
            }

            if (plaettchen == 16 || plaettchen == 17)
            {
                pieces.Remove(16);
                pieces.Remove(17);

                return;
            }

            pieces.Remove(plaettchen);
        }

        private static LinkedList<int> GetAvailablePieces(IcoSoku ico)
        {
            var pieces = new LinkedList<int>();

            for (var plaettchen = 0; plaettchen < 20; plaettchen++)
            {
                if (ico.IstPlaettchenVerfuegbar(plaettchen))
                {
                    pieces.AddLast(plaettchen);
                }
            }

            return pieces;
        }
    }
}